docs/design/map/map-store-operation-offloading.md

```diff
@@ -0,0 +1,247 @@
+          --------------------------------------------------------      |
+                                  |                                     | 
+                                  |                                     |
+                                  |                                     |
+          ----------------------------------------------------    If No |
+                  Do we encounter offload needed Step?      ------------+
+          ----------------------------------------------------
+                                  | 
+                                  | If Yes
+                                  |
+          -------------------------------------------------------
+                  Offload next Step and free partition thread
+          -------------------------------------------------------
+                                  | 
+                                  |
+                          --------------------
+                            End of execution
+                          --------------------
+
+
+#### Flow in Offloaded Thread
+          -------------------------------------------------------
+            Run offload needed Step of the operation here         <-----+
+          --------------------------------------------------------      |
+                                  |                                     | 
+                                  |                                     |
+          ----------------------------------------------------   If Yes |
+                  Do we encounter offload needed Step?      ------------+
+          ----------------------------------------------------
+                                  | 
+                                  | If No
+                                  |
+          -------------------------------------------------------
+              Schedule next Step to run in partition thread
+          -------------------------------------------------------
+                                  | 
+                                  |
+                          --------------------
+                            End of execution
+                          --------------------
+
+### Risks and Assumptions
+All MapStore interacting operations are queued per map per
+partition and next operations must wait end of head-operation
+before execution, although this is an improvement when we
+compare it with the current situation(which is all operations
+of all maps in a partition waiting head-operation), it can
+also be deemed as a limitation to preserve current behavior.
+
+**NOTE**: IMap api calls like `getEntryView` will not be queued
+even map has a map-store, since it only returns in memory
+view of an entry, and it has no map-store interaction, it is ok
+to run this kind of api calls without waiting previously offloaded operations.
+What this means is can be seen as in below example:
+
+```java
+map.put
+map.getEntryView --> always sees result of map.putAsync
```

Comment:
Yes, it has same guarantee with current behavior for sync `put`. But only `putAsync` can be not-observed since we have currently an implicit order there for the same key.

Reply:
Good, can you please clarify this in the text above? How do we ensure that the results of a `map.put` are visible in `getEntryView` if the former is queued but the latter is not?

Reply:
done